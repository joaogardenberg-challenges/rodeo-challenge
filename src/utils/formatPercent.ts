export default function formatPercent(price: number) {
  return price.toLocaleString('en', {
    minimumFractionDigits: 0,
    maximumFractionDigits: 2
  })
}
