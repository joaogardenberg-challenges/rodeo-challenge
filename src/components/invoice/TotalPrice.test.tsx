import { render, screen } from '@testing-library/react'
import { noop } from 'lodash'
import formatPercent from 'utils/formatPercent'
import formatPrice from 'utils/formatPrice'

const useContext = jest.fn()
const getInvoicePrice = jest.fn()
const getInvoicePriceWithFeeOrDiscountPercent = jest.fn()
const getInvoicePriceWithTax = jest.fn()

jest.doMock('react', () => ({ useContext }))

jest.doMock('utils/invoicePrice', () => ({
  getInvoicePrice,
  getInvoicePriceWithFeeOrDiscountPercent,
  getInvoicePriceWithTax
}))

jest.doMock('./Invoice', () => ({ InvoiceContext: noop }))

const TotalPrice = require('./TotalPrice').default

const price = 10
const fee = 6
const discount = -6
const tax = 2

const invoices = {
  none: { currency: 'USD' },
  fee: { currency: 'USD', feeOrDiscountPercent: fee },
  discount: { currency: 'USD', feeOrDiscountPercent: discount }
}

describe('TotalPrice', () => {
  beforeEach(() => {
    getInvoicePrice.mockReturnValue(price)
  })

  it('Renders successfully without feeOrDiscountPercent', () => {
    useContext.mockReturnValue(invoices.none)
    getInvoicePriceWithFeeOrDiscountPercent.mockReturnValue(price)
    getInvoicePriceWithTax.mockReturnValue(price + tax)

    const { container } = render(<TotalPrice />)
    const subtotalPrice = screen.getByTestId('subtotal-price')
    const feeOrDiscountLabel = screen.queryByTestId('fee-or-discount-label')
    const feeOrDiscountAmount = screen.queryByTestId('fee-or-discount-amount')
    const feeOrDiscountPrice = screen.queryByTestId('fee-or-discount-price')
    const taxAmount = screen.getByTestId('tax-amount')
    const totalPrice = screen.getByTestId('total-price')

    expect(subtotalPrice).toHaveTextContent(`$${formatPrice(price)}`)
    expect(feeOrDiscountLabel).not.toBeInTheDocument()
    expect(feeOrDiscountAmount).not.toBeInTheDocument()
    expect(feeOrDiscountPrice).not.toBeInTheDocument()
    expect(taxAmount).toHaveTextContent(`$${formatPrice(tax)}`)
    expect(totalPrice).toHaveTextContent(`$${formatPrice(price + tax)}`)
    expect(container).toMatchSnapshot()
  })

  it('Renders successfully with feePercent', () => {
    useContext.mockReturnValue(invoices.fee)
    getInvoicePriceWithFeeOrDiscountPercent.mockReturnValue(price + fee)
    getInvoicePriceWithTax.mockReturnValue(price + fee + tax)

    const { container } = render(<TotalPrice />)
    const subtotalPrice = screen.getByTestId('subtotal-price')
    const feeOrDiscountLabel = screen.getByTestId('fee-or-discount-label')
    const feeOrDiscountAmount = screen.getByTestId('fee-or-discount-amount')
    const feeOrDiscountPrice = screen.getByTestId('fee-or-discount-price')
    const taxAmount = screen.getByTestId('tax-amount')
    const totalPrice = screen.getByTestId('total-price')

    expect(subtotalPrice).toHaveTextContent(`$${formatPrice(price)}`)

    expect(feeOrDiscountLabel).toHaveTextContent(
      `Subtotal price with fee (+ ${formatPercent(fee)}%)`
    )

    expect(feeOrDiscountAmount).toHaveTextContent(`+ $${formatPrice(fee)}`)
    expect(feeOrDiscountPrice).toHaveTextContent(`$${formatPrice(price + fee)}`)
    expect(taxAmount).toHaveTextContent(`$${formatPrice(tax)}`)
    expect(totalPrice).toHaveTextContent(`$${formatPrice(price + fee + tax)}`)
    expect(container).toMatchSnapshot()
  })

  it('Renders successfully with discountPercent', () => {
    useContext.mockReturnValue(invoices.discount)
    getInvoicePriceWithFeeOrDiscountPercent.mockReturnValue(price + discount)
    getInvoicePriceWithTax.mockReturnValue(price + discount + tax)

    const { container } = render(<TotalPrice />)
    const subtotalPrice = screen.getByTestId('subtotal-price')
    const feeOrDiscountLabel = screen.getByTestId('fee-or-discount-label')
    const feeOrDiscountAmount = screen.getByTestId('fee-or-discount-amount')
    const feeOrDiscountPrice = screen.getByTestId('fee-or-discount-price')
    const taxAmount = screen.getByTestId('tax-amount')
    const totalPrice = screen.getByTestId('total-price')

    expect(subtotalPrice).toHaveTextContent(`$${formatPrice(price)}`)

    expect(feeOrDiscountLabel).toHaveTextContent(
      `Subtotal price with discount (- ${formatPercent(Math.abs(discount))}%)`
    )

    expect(feeOrDiscountAmount).toHaveTextContent(
      `- $${formatPrice(Math.abs(discount))}`
    )

    expect(feeOrDiscountPrice).toHaveTextContent(
      `$${formatPrice(price + discount)}`
    )

    expect(taxAmount).toHaveTextContent(`$${formatPrice(tax)}`)

    expect(totalPrice).toHaveTextContent(
      `$${formatPrice(price + discount + tax)}`
    )

    expect(container).toMatchSnapshot()
  })
})
