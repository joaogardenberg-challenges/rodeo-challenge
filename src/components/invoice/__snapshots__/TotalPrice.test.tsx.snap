// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TotalPrice Renders successfully with discountPercent 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-size: 1.2rem;
  gap: 1rem;
  padding: 3rem 1rem;
}

.c0 .total-title {
  font-size: 2rem;
  text-align: center;
}

<div>
  <section
    class="c0 total-price-section"
  >
    <h2
      class="total-title"
    >
      Total
    </h2>
    <ol
      class="total-prices"
    >
      <li
        aria-hidden="true"
        class="cost-item cost-item--labels"
      >
        <span
          class="cost-item__name"
        />
        <span
          class="cost-item__tax"
        />
        <span
          class="cost-item__description"
        >
          Description
        </span>
        <span
          class="cost-item__price"
        >
          Price
        </span>
      </li>
      <li
        class="cost-item cost-item--total"
      >
        <span
          class="cost-item__name"
        >
          Subtotal price
        </span>
        <span
          hidden=""
        >
          : 
        </span>
        <span
          class="cost-item__tax"
        />
        <span
          class="cost-item__description"
        />
        <span
          class="cost-item__price"
          data-testid="subtotal-price"
        >
          $
          10.00
        </span>
      </li>
      <li
        class="cost-item cost-item--total"
      >
        <span
          class="cost-item__name"
          data-testid="fee-or-discount-label"
        >
          Subtotal price with 
          discount
           
          <span
            class="cost-item__number"
          >
            (
            -
             
            6
            %)
          </span>
        </span>
        <span
          hidden=""
        >
          : 
        </span>
        <span
          class="cost-item__tax"
        />
        <span
          class="cost-item__description"
          data-testid="fee-or-discount-amount"
        >
          -
           
          $
          6.00
        </span>
        <span
          hidden=""
        >
           = 
        </span>
        <span
          class="cost-item__price"
          data-testid="fee-or-discount-price"
        >
          $
          4.00
        </span>
      </li>
      <li
        class="cost-item cost-item--total"
      >
        <span
          class="cost-item__name"
        >
          Total price with tax
        </span>
        <span
          hidden=""
        >
          : 
        </span>
        <span
          class="cost-item__tax"
        />
        <span
          class="cost-item__description"
          data-testid="tax-amount"
        >
          + 
          $
          2.00
        </span>
        <span
          hidden=""
        >
           = 
        </span>
        <span
          class="cost-item__price"
          data-testid="total-price"
        >
          $
          6.00
        </span>
      </li>
    </ol>
  </section>
</div>
`;

exports[`TotalPrice Renders successfully with feePercent 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-size: 1.2rem;
  gap: 1rem;
  padding: 3rem 1rem;
}

.c0 .total-title {
  font-size: 2rem;
  text-align: center;
}

<div>
  <section
    class="c0 total-price-section"
  >
    <h2
      class="total-title"
    >
      Total
    </h2>
    <ol
      class="total-prices"
    >
      <li
        aria-hidden="true"
        class="cost-item cost-item--labels"
      >
        <span
          class="cost-item__name"
        />
        <span
          class="cost-item__tax"
        />
        <span
          class="cost-item__description"
        >
          Description
        </span>
        <span
          class="cost-item__price"
        >
          Price
        </span>
      </li>
      <li
        class="cost-item cost-item--total"
      >
        <span
          class="cost-item__name"
        >
          Subtotal price
        </span>
        <span
          hidden=""
        >
          : 
        </span>
        <span
          class="cost-item__tax"
        />
        <span
          class="cost-item__description"
        />
        <span
          class="cost-item__price"
          data-testid="subtotal-price"
        >
          $
          10.00
        </span>
      </li>
      <li
        class="cost-item cost-item--total"
      >
        <span
          class="cost-item__name"
          data-testid="fee-or-discount-label"
        >
          Subtotal price with 
          fee
           
          <span
            class="cost-item__number"
          >
            (
            +
             
            6
            %)
          </span>
        </span>
        <span
          hidden=""
        >
          : 
        </span>
        <span
          class="cost-item__tax"
        />
        <span
          class="cost-item__description"
          data-testid="fee-or-discount-amount"
        >
          +
           
          $
          6.00
        </span>
        <span
          hidden=""
        >
           = 
        </span>
        <span
          class="cost-item__price"
          data-testid="fee-or-discount-price"
        >
          $
          16.00
        </span>
      </li>
      <li
        class="cost-item cost-item--total"
      >
        <span
          class="cost-item__name"
        >
          Total price with tax
        </span>
        <span
          hidden=""
        >
          : 
        </span>
        <span
          class="cost-item__tax"
        />
        <span
          class="cost-item__description"
          data-testid="tax-amount"
        >
          + 
          $
          2.00
        </span>
        <span
          hidden=""
        >
           = 
        </span>
        <span
          class="cost-item__price"
          data-testid="total-price"
        >
          $
          18.00
        </span>
      </li>
    </ol>
  </section>
</div>
`;

exports[`TotalPrice Renders successfully without feeOrDiscountPercent 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-size: 1.2rem;
  gap: 1rem;
  padding: 3rem 1rem;
}

.c0 .total-title {
  font-size: 2rem;
  text-align: center;
}

<div>
  <section
    class="c0 total-price-section"
  >
    <h2
      class="total-title"
    >
      Total
    </h2>
    <ol
      class="total-prices"
    >
      <li
        aria-hidden="true"
        class="cost-item cost-item--labels"
      >
        <span
          class="cost-item__name"
        />
        <span
          class="cost-item__tax"
        />
        <span
          class="cost-item__description"
        >
          Description
        </span>
        <span
          class="cost-item__price"
        >
          Price
        </span>
      </li>
      <li
        class="cost-item cost-item--total"
      >
        <span
          class="cost-item__name"
        >
          Subtotal price
        </span>
        <span
          hidden=""
        >
          : 
        </span>
        <span
          class="cost-item__tax"
        />
        <span
          class="cost-item__description"
        />
        <span
          class="cost-item__price"
          data-testid="subtotal-price"
        >
          $
          10.00
        </span>
      </li>
      <li
        class="cost-item cost-item--total"
      >
        <span
          class="cost-item__name"
        >
          Total price with tax
        </span>
        <span
          hidden=""
        >
          : 
        </span>
        <span
          class="cost-item__tax"
        />
        <span
          class="cost-item__description"
          data-testid="tax-amount"
        >
          + 
          $
          2.00
        </span>
        <span
          hidden=""
        >
           = 
        </span>
        <span
          class="cost-item__price"
          data-testid="total-price"
        >
          $
          12.00
        </span>
      </li>
    </ol>
  </section>
</div>
`;
